src/storage/src/dedup_pk_cell_based_row_serializer.rs

```diff
@@ -0,0 +1,204 @@
+use risingwave_common::error::Result;
+use risingwave_common::types::VirtualNode;
+
+use crate::cell_based_row_serializer::CellBasedRowSerializer;
+use crate::cell_serializer::{CellSerializer, KeyBytes, ValueBytes};
+
+/// [`DedupPkCellBasedRowSerializer`] is identical to [`CellBasedRowSerializer`].
+/// Difference is that before serializing a row, pk datums are filtered out.
+pub struct DedupPkCellBasedRowSerializer {
+    /// Contains:
+    /// 1. Row indices of datums not in pk,
+    /// 2. or datums which have to be stored regardless
+    ///    (e.g. if memcomparable not equal to value encoding)
+    dedup_datum_indices: HashSet<usize>,
+
+    /// Serializing of row after filtering pk datums
+    /// should be same as `CellBasedRowSerializer`.
+    /// Hence we reuse its functionality.
+    inner: CellBasedRowSerializer,
+}
+
+impl DedupPkCellBasedRowSerializer {
+    /// Constructs a new [`DedupPkCellBasedRowSerializer`].
+    pub fn new(pk_indices: &[usize], column_descs: &[ColumnDesc], column_ids: &[ColumnId]) -> Self {
+        let pk_indices = pk_indices.iter().cloned().collect::<HashSet<_>>();
+        let dedup_datum_indices = (0..column_descs.len())
+            .filter(|i| {
+                !pk_indices.contains(i) || !column_descs[*i].data_type.mem_cmp_eq_value_enc()
+            })
+            .collect();
+        let dedupped_column_ids = Self::remove_dup_pk_column_ids(&dedup_datum_indices, column_ids);
+        let inner = CellBasedRowSerializer::new(dedupped_column_ids);
+        Self {
+            dedup_datum_indices,
+            inner,
+        }
+    }
+
+    /// Used internally to filter through an iterator,
+    /// finding items which should be in dedup pk row.
+    fn filter_by_dedup_datum_indices<'b, I>(
+        dedup_datum_indices: &'b HashSet<usize>,
+        iter: impl Iterator<Item = I> + 'b,
+    ) -> impl Iterator<Item = I> + 'b {
+        iter.enumerate()
+            .filter(|(i, _)| dedup_datum_indices.contains(i))
+            .map(|(_, d)| d)
+    }
+
+    /// Filters out duplicate pk datums by reference.
+    #[allow(dead_code)]
+    fn remove_dup_pk_datums_by_ref(&self, row: &Row) -> Row {
+        Row(
+            Self::filter_by_dedup_datum_indices(&self.dedup_datum_indices, row.0.iter())
+                .cloned()
+                .collect(),
+        )
+    }
+
+    /// Filters out duplicate pk datums.
+    fn remove_dup_pk_datums(&self, row: Row) -> Row {
+        Row(
+            Self::filter_by_dedup_datum_indices(&self.dedup_datum_indices, row.0.into_iter())
+                .collect(),
+        )
+    }
+
+    /// Filters out column ids duplicate
+    fn remove_dup_pk_column_ids(
+        dedup_datum_indices: &HashSet<usize>,
+        column_ids: &[ColumnId],
+    ) -> Vec<ColumnId> {
+        Self::filter_by_dedup_datum_indices(dedup_datum_indices, column_ids.iter())
+            .cloned()
+            .collect()
+    }
+}
+
+impl CellSerializer for DedupPkCellBasedRowSerializer {
```

Comment:
The name looks a bit confusing for me. 
`RowSerializer` seems serialize a row (vec<Datum>), while `CellSerializer` for a cell (Datum).